library/Booster/Pattern/UnifiedMatcher.hs

```diff
@@ -0,0 +1,839 @@
+                    (True, False) ->
+                        -- subject has extra assocs to match with pattern rest
+                        let subj' = KMap subjDef (Map.assocs subjExtra) subjRest
+                         in case patRest of
+                                Nothing ->
+                                    failWith $ DifferentValues emptyMap subj'
+                                Just pRest -> do
+                                    enqueueRegularProblem pRest subj'
+                    (False, True) ->
+                        -- pattern has extra assocs
+                        let pat' = KMap patDef (Map.assocs patExtra) patRest
+                         in case subjRest of
+                                Nothing ->
+                                    -- no opaque rest, match is definitely failing
+                                    failWith $ DifferentValues pat' emptyMap
+                                Just sRest ->
+                                    -- indeterminate matching with an opaque rest
+                                    addIndeterminate pat' sRest
+                    (False, False)
+                        -- Special case: definitely fail if all (extra) pattern keys are concrete
+                        -- and there is no opaque subject rest
+                        | Nothing <- subjRest
+                        , all isConcrete (Map.keys patExtra) ->
+                            let pat' = KMap patDef (Map.assocs patExtra) patRest
+                                subj' = KMap subjDef (Map.assocs subjExtra) subjRest
+                             in failWith $ DifferentValues pat' subj'
+                        -- Special case: attempt a match if pattern and subject assocs
+                        -- are singleton and there is no opaque subject rest
+                        | Nothing <- subjRest
+                        , [(pKey, pVal)] <- Map.assocs patExtra
+                        , [(sKey, sVal)] <- Map.assocs subjExtra -> do
+                            let opaque = case patRest of
+                                    Nothing -> []
+                                    Just rest -> [(rest, emptyMap)]
+                            enqueueRegularProblems . Seq.fromList $ (pKey, sKey) : (pVal, sVal) : opaque
+                        | otherwise ->
+                            addIndeterminate t1 t2
+
+                -- enqueue common elements for matching unless already failed
+                enqueueRegularProblems $ Seq.fromList $ Map.elems commonMap
+      where
+        emptyMap = KMap patDef [] Nothing
+
+        checkDuplicateKeys m@(KMap _ assocs _) =
+            let duplicates =
+                    Map.filter (> (1 :: Int)) $ foldr (flip (Map.insertWith (+)) 1 . fst) mempty assocs
+             in case Map.assocs duplicates of
+                    [] -> pure ()
+                    (k, _) : _ -> failWith $ DuplicateKeys k m
+        checkDuplicateKeys _ = pure ()
+matchMaps _ _ _ = undefined
```

Comment:
This is probably an overlooked case, or should it simple "never happen"? Either way, probably a good idea to return indeterminate here.

Reply:
this is getting a refactor in the followup PR anyway